### DISCUSSION TITLE
Facebook provider authorization params

### DISCUSSION THREAD

**Discussion by gkielwasser** [nextauthjs/next-auth#2807] [Help] [answered]
Hi,

It seems I cannot add any parameters for the Facebook provider.

Concerning the Google provider i did the following and the prompt parameter is well append to the url.
```
providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
      authorization: { params: { prompt: 'consent' } },
    }),
  ],
```

I tried the same for Facebook but the display parameter is not appended in the url.
```
providers: [
    FacebookProvider({
      clientId: process.env.FACEBOOK_ID,
      clientSecret: process.env.FACEBOOK_SECRET,
      authorization: { params: { display: 'popup' } }
    })
  ],
```

I'm using the following Nextauth version ``` 4.0.0-beta.2 ```

Thanks

**Comment by balazsorban44** [comment, marked as answer]
It might be that we simply don't support `display: popup`. When we created the authorization URL, we redirect the user to that URL.

But a quick google search gave me this:

> You need to include the Facebook JS SDK for those popups to work correctly. Instructions on how to include the SDK are available in the developer documentation.

https://stackoverflow.com/a/10179342/5364135

so maybe start there and see if that helps.

**Reply by gkielwasser** [reply]
> we redirect the user to that URL

Why since 4.0.0 beta we cannot anymore configure anything in the Facebook authorization url ?

[This helped ](https://github.com/nextauthjs/next-auth/blob/245567bb98adbc7d918b535b311e990d209cfcdf/src/server/lib/oauth/authorization-url.js#L22-L30)me to understand why it is dropped

Thanks

**Reply by balazsorban44** [reply]
not sure what you mean. if anything, we expanded the things you could configure. 👀

**Reply by gkielwasser** [reply]
What I understand from the quoted code is that if the provider specify an "authorization" that is a string, it will enforce to use this provider with the parameters specified in the url.
Otherwise if we want to let the "authorization" params configurable, the base provider must have an "authorization" property that is not a string.

So I'm not sure why we enforce not to be able to configure the url for Facebook :). (Previously, it was possible because [the options was spread](https://github.com/nextauthjs/next-auth/blob/ead715219a5d7a6e882a6ba27fa56b03954d062d/src/providers/facebook.js#L20))

I didn't find an exhaustive list of the parameters but at least the "display" option exist [here](https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow/)

**Reply by balazsorban44** [reply]
You can override the `authorization` property. Maybe these would help:

https://github.com/nextauthjs/next-auth/pull/2411#issue-693918157
https://next-auth.js.org/configuration/providers/oauth-provider#options

**Reply by gkielwasser** [reply]
Sorry I cannot follow you. Reading both of your 2 links, what I state in my first comment looks like the way I should override the authorization, isn't it ?

```
import FacebookProvider  from "next-auth/providers/facebook";

export default function NextAuth({
  ...,
  providers: [
     FacebookProvider({
      clientId: process.env.FACEBOOK_ID,
      clientSecret: process.env.FACEBOOK_SECRET,
      authorization: { params: { display: 'popup' } }
    })
  ]
})
```

I would say there is something wrong with the [merge function](https://github.com/nextauthjs/next-auth/blob/17bea4ab69ac3dfbf4e8d2146cedaa62439c7f18/src/lib/merge.ts#L16). If we try to override an authorization that is a string with an authorization that is an object, it won't assign the user options.

To clarify what happen during the merge:
the given input ``` "authorization" : "anAuthorizationUrlString" ``` woudn't be overrided with ``` "authorization": { params: { } }```.
The merge output will still be ``` "authorization" : "anAuthorizationUrlString" ```

**Reply by balazsorban44** [reply]
I've been going around thinking of this why I did not understand your problem. I finally think I see the issue, and it comes from the following place:

https://github.com/nextauthjs/next-auth/blob/506672676a23aed6f50868b902621af1d0ec6526/src/lib/merge.ts#L13

We use a `merge` function to merge the user's config options with the default one, and it only does merging if both the target and source are objects. We should normalize the incoming configs before merging to avoid this problem.

I came across this yesterday myself when I wanted to override the SpotifyProvider, which had the same string formatted default `authorization`: https://github.com/nextauthjs/next-auth/blob/506672676a23aed6f50868b902621af1d0ec6526/src/providers/spotify.js#L8

I'm sorry that it took so long to realize the problem here :sweat_smile: 

I think https://github.com/nextauthjs/next-auth/blob/beta/src/server/lib/providers.ts is a good place to parse strings to objects before merging. Will open a PR soon.

**Reply by balazsorban44** [reply]
@gkielwasser I am addressing this in #2842!

If you want to try it out, you can install it like so `npm i next-auth@0.0.0-pr.2842.ccf36f15`

UPDATE: It's merged, and will be available in the next `beta` release, no ETA yet.

**Reply by gkielwasser** [reply]
@balazsorban44  Yep it is exactly the part of code I suspected :)
Thank you very much for that fix, I will test it as soon as possible!
